csrc/core/scalar_type.hpp

```diff
@@ -0,0 +1,368 @@
+                  "Cannot represent max as a int64_t");
+      return {(int64_t(1) << mantissa) - 1};
+    }
+  }
+
+  std::variant<int64_t, double> _raw_min() const {
+    if (is_floating_point()) {
+      TORCH_CHECK(is_signed(),
+                  "We currently assume all floating point types are signed");
+      constexpr uint64_t sign_bit_double = (uint64_t(1) << 63);
+
+      double max = _floating_point_max();
+      uint64_t max_raw = *reinterpret_cast<uint64_t*>(&max);
+      uint64_t min_raw = max_raw | sign_bit_double;
+      return {*reinterpret_cast<double*>(&min_raw)};
+    } else {
+      TORCH_CHECK(!is_signed() || size_bits() <= 64,
+                  "Cannot represent min as a int64_t");
+      if (is_signed()) {
+        // set the top bit to 1 (i.e. INT64_MIN) and the rest to 0
+        // then perform an arithmetic shift right to set all the bits above
+        // (size_bits() - 1) to 1
+        return {INT64_MIN >> (64 - size_bits())};
+      } else {
+        return {int64_t(0)};
+      }
+    }
+  }
+
+ public:
+  // Max representable value for this scalar type.
+  // (accounting for bias if there is one)
+  std::variant<int64_t, double> max() const {
+    return std::visit(
+        [this](auto x) -> std::variant<int64_t, double> { return {x - bias}; },
+        _raw_max());
+  }
+
+  // Min representable value for this scalar type.
+  // (accounting for bias if there is one)
+  std::variant<int64_t, double> min() const {
+    return std::visit(
+        [this](auto x) -> std::variant<int64_t, double> { return {x - bias}; },
+        _raw_min());
+  }
+
+  std::string str() const {
+    /*
+     * generally follows: https://github.com/jax-ml/ml_dtypes
+     * for floating point types (leading f):
+     *  - E_: exponent size
+     *  - M_: mantissa size
+     *  - no-trailing letters: means it follows IEEE 754 conventions
+     *  - trailing f: means finite values only (no infinities)
+     *  - trailing n: means nans are supported (non-standard encoding)
+     * for integer types (leading s/u):
+     *  - leading s: means signed
+     *  - leading u: means unsigned
+     *  - number following s/u: number of bits
+     *  - bX: indicates a non-zero bias of X
+     */
+    if (is_floating_point()) {
+      auto ret =
+          "fE" + std::to_string(exponent) + "M" + std::to_string(mantissa);
+      if (!is_ieee_754()) {
+        if (finite_values_only) {
+          ret += "f";
+        }
+        if (nan_repr != NAN_NONE) {
+          ret += "n";
+        }
+      }
+      return ret;
+    } else {
+      auto ret = ((is_signed()) ? "s" : "u") + std::to_string(size_bits());
+      if (has_bias()) {
+        ret += "z" + std::to_string(bias);
+      }
+      return ret;
+    }
+  }
+
+  bool operator==(ScalarType const& other) const {
+    return mantissa == other.mantissa && exponent == other.exponent &&
+           bias == other.bias && _signed == other._signed;
```

Comment:
should this also be checking the nan + finite members?

Reply:
good catch! thanks!